res/controllers/Behringer-CMDStudio4a-scripts.js

```diff
@@ -0,0 +1,257 @@
+}
+
+BehringerCMDStudio4a.init = function () {
+    // Initialise anything that might not be in the correct state.
+    BehringerCMDStudio4a.initLEDs();
+    // Connect the VUMeters
+    engine.connectControl("[Master]","VuMeterL","BehringerCMDStudio4a.vuMeterUpdate");
+    engine.connectControl("[Master]","VuMeterR","BehringerCMDStudio4a.vuMeterUpdate");
+}
+ 
+BehringerCMDStudio4a.shutdown = function() {
+    // Leave the deck in a properly initialised state.
+    BehringerCMDStudio4a.initLEDs();
+
+    // Disconnect the VUMeters.
+    // Maybe not! This doesn't seem to work here!
+//  engine.connectControl("[Master]","VuMeterL","BehringerCMDStudio4a.vuMeterUpdate",true);
+//  engine.connectControl("[Master]","VuMeterR","BehringerCMDStudio4a.vuMeterUpdate",true);
+}
+
+
+// *************************** Control Stuff. *********************************
+// The code below is primarily "shift/mode" key functionality as there is no
+// native support for this in Mixxx at the moment.
+// I suspect that the vast majority of controller mappings could be completed
+// with little or no scripting if Mixxx supported shift/mode buttons in the
+// XML (together with standard wheel/scratching functionality).
+
+// Function to deal with the DEL "shift" buttons.
+BehringerCMDStudio4a.del = function (channel, control, value, status, group) {
+    BehringerCMDStudio4a.delState[channel]=!BehringerCMDStudio4a.delState[channel];
+    midi.sendShortMsg(status, control, BehringerCMDStudio4a.delState[channel] ? 0x01 : 0x00);
+}
+
+// Functions to deal with the play button.
+BehringerCMDStudio4a.playPush = function (channel, control, value, status, group) {
+    if (BehringerCMDStudio4a.delState[channel]) {
+        // DEL is active, reverse-roll (slip)!
+        engine.setValue(group, "reverseroll", 1);
+    } else {
+        // No DEL, just play.
+        if (engine.getValue(group, "play")==0) {
+            engine.setValue(group, "play", 1);
+        } else {
+            engine.setValue(group, "play", 0);
+        }
+    }
+}
+BehringerCMDStudio4a.playRelease = function (channel, control, value, status, group) {
+    if (BehringerCMDStudio4a.delState[channel]) {
+        // DEL is active, reverse-roll (slip) release!
+        engine.setValue(group, "reverseroll", 0);
+    }
+}
+
+// Functions to deal with the cue button.
+BehringerCMDStudio4a.cuePush = function (channel, control, value, status, group) {
+    if (BehringerCMDStudio4a.delState[channel]) {
+        // DEL is active, reverse play!
+        engine.setValue(group, "reverse", 1);
+    } else {
+        // No DEL so just cue.
+        engine.setValue(group, "cue_default", 1);
+    }
+}
+BehringerCMDStudio4a.cueRelease = function (channel, control, value, status, group) {
+    if (BehringerCMDStudio4a.delState[channel]) {
+        // DEL is active, reverse-roll release!
+        engine.setValue(group, "reverse", 0);
+    } else {
+        engine.setValue(group, "cue_default", 0);
+	}
+}

```

Comment:
See earlier comment.
I've only done this where the function is different for each (press or release). I didn't know about the toggleControl()  function at the time, so I could replace the cuePush() and cueRelease() functions with a cue() "super-function" using toggleControl() but thinking about it now, what if the behaviour get's out of sync for some reason (e.g. if the GUI is touched perhaps?, ,or because of some other odd interaction??). If that happens the Push/Release functions would get swapped, (with no simple way for the user to restore normal behaviour). Perhaps this is unlikely, but I believe in "defensive programming" :-)

Perhaps you could get round this by keeping the explicit setValue() calls and wrapping the super function in an if condition to detect whether we are processing a push or a release, but then we are back to adding an if clause to the "super-function" to do this (i.e. creating a more complex function) when we already have the ability to do this test in the XML.

If single functions for both press and release is the Mixxx "house-style" then I will change this (let me know), but (IMHO) it seems simpler the way it is (although again, if shift/mode functionality could be declared directly in the XML without additional scripting, life would be much easier" :-))